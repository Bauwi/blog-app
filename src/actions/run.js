import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
