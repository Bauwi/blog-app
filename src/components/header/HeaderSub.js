import React from 'react';
import { NavLink } from 'react-router-dom';

const HeaderSub = () => (
  <nav className="header__navbarsub">
    <div className="header-navbarsub--three">
      <NavLink to="/category/music" activeClassName="header__navbarsub__tab--selected">
        Music
      </NavLink>
      <NavLink to="/category/movie" activeClassName="header__navbarsub__tab--selected">
        Movie
      </NavLink>
      <NavLink to="/category/litterature" activeClassName="header__navbarsub__tab--selected">
        Litterature
      </NavLink>
    </div>
    <div className="header-navbarsub--three">
      <NavLink to="/category/life" activeClassName="header__navbarsub__tab--selected">
        Life
      </NavLink>
      <NavLink to="/category/programming" activeClassName="header__navbarsub__tab--selected">
        Programming
      </NavLink>
      <NavLink to="/category/history" activeClassName="header__navbarsub__tab--selected">
        History
      </NavLink>
    </div>
    <div className="header-navbarsub--three">
      <NavLink to="/category/sport" activeClassName="header__navbarsub__tab--selected">
        Sport
      </NavLink>
      <NavLink to="/category/business" activeClassName="header__navbarsub__tab--selected">
        Business
      </NavLink>
    </div>
  </nav>
);

export default HeaderSub;
