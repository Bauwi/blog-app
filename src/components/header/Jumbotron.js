import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import PublicPostList from '../PublicPostsList';
import { Carousel } from 'antd';

import JumbotronSlide from './JumbotronSlide';

import selectCarouselPosts from '../../selectors/carousel';

export class Jumbotron extends Component {
  state = {
    currentSlide: 1
  };
  renderSlides = slideIndex => {
    const post = this.props.posts[slideIndex];
    return <JumbotronSlide post={post} />;
  };
  handleClick = () => {
    const pathname = `${this.props.posts[this.state.currentSlide - 1].authorId}/read/${this.props
      .posts[this.state.currentSlide - 1].id}`;
    this.props.history.push(pathname);
  };
  render() {
    if (this.props.isLoading || this.props.posts.length === 0) {
      return <p />;
    }
    return (
      <div className="carousel">
        <Carousel
          effect="fade"
          autoplay
          autoplaySpeed={7000}
          afterChange={currentSlide => {
            if (this.state.currentSlide === 6) {
              return this.setState(() => ({ currentSlide: 1 }));
            }
            this.setState({ currentSlide: currentSlide + 1 });
          }}
        >
          <div onClick={this.handleClick}>{this.renderSlides(0)}</div>
          <div onClick={this.handleClick}>{this.renderSlides(1)}</div>
          <div onClick={this.handleClick}>{this.renderSlides(2)}</div>
          <div onClick={this.handleClick}>{this.renderSlides(3)}</div>
          <div onClick={this.handleClick}>{this.renderSlides(4)}</div>
          <div onClick={this.handleClick}>{this.renderSlides(5)}</div>
        </Carousel>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  posts: selectCarouselPosts(
    state.readings.posts,
    !!state.users.preferences
      ? state.users.preferences.topCategories
      : ['music', 'literature', 'cinema']
  ),
  isLoading: state.readings.isLoading
});

export default withRouter(connect(mapStateToProps)(Jumbotron));
