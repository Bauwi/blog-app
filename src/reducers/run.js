const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
