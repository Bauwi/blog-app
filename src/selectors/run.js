const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
