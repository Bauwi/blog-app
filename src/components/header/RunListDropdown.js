import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Menu, Dropdown, Icon, message } from 'antd';

import { setCurrentPostRun } from '../../actions/run';

const warning = () => {
  message.warning('You have already read all the posts contained in your run.');
};

export class RunListDropdown extends Component {
  renderDropdownPosts = () => {
    return this.props.posts.map(post => {
      return (
        <Menu.Item key={post.content.id}>
          <div
            className="dropdown-line"
            onClick={() => {
              this.props.setCurrentPostRun(post.content.id);
              this.props.history.push('/run/start');
            }}
          >
            <img src={post.content.miniCover} alt="minicover" />
            <p>{post.content.title}</p>
          </div>
        </Menu.Item>
      );
    });
  };

  handleResumeClick = () => {
    const firstUnreadIndex = this.props.posts.findIndex(post => post.state === 'unread');
    if (firstUnreadIndex === -1) {
      warning();
      this.props.setCurrentPostRun(this.props.posts[0].content.id);
      return this.props.history.push('/run');
    }
    this.props.setCurrentPostRun(this.props.posts[firstUnreadIndex].content.id);
    return this.props.history.push('/run/start');
  };

  render() {
    console.log(this.props);
    const menu = (
      <Menu>
        <Menu.Item key="0">
          <button className="button" onClick={this.handleResumeClick}>
            Resume
          </button>
          <button className="button">Reset</button>
        </Menu.Item>
        {this.props.posts && this.renderDropdownPosts()}
      </Menu>
    );
    return (
      <Dropdown overlay={menu} placement="bottomCenter" className="dropdown">
        <Link className="ant-dropdown-link" to="/run">
          <i className="fa fa-book" />
        </Link>
      </Dropdown>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setCurrentPostRun: id => dispatch(setCurrentPostRun(id))
});

const mapStateToProps = state => ({
  posts: state.run.posts
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(RunListDropdown));
