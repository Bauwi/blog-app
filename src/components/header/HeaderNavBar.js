import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { Badge } from 'antd';

import RunListDropdown from './RunListDropdown';
import LoginModal from './LoginModal';

import { startLogout } from '../../actions/auth';
import { startSetRunPosts } from '../../actions/run';

export class HeaderNavBar extends Component {
  componentDidMount() {
    if (this.props.isAuthenticated) {
      this.props.startSetRunPosts();
    }
  }

  render() {
    const { isAuthenticated, runCount = 0, startLogout } = this.props;
    if (isAuthenticated) {
      return (
        <nav className="header__navbar">
          <NavLink to="/">
            <i className="fa fa-home" />
          </NavLink>
          <NavLink to="/dashboard">
            <i className="fa fa-pencil" />
          </NavLink>
          <nav to="/run">
            <Badge className="header__navbar__run__badge" showZero count={runCount} />
            <RunListDropdown />
          </nav>
          <NavLink to="/preferences">
            <i className="fa fa-cog" />
          </NavLink>
          <button className="button button--navbar" onClick={startLogout}>
            <i className="fa fa-sign-out" />
          </button>
        </nav>
      );
    }
    return (
      <nav className="header__navbar">
        <NavLink to="/">
          <i className="fa fa-home" />
        </NavLink>
        <LoginModal />
      </nav>
    );
  }
}

HeaderNavBar.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  runCount: PropTypes.number,
  startLogout: PropTypes.func.isRequired,
  startSetRunPosts: PropTypes.func.isRequired
};

HeaderNavBar.defaultProps = { runCount: 0 };

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout()),
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.uid,
  runCount: state.run.posts ? state.run.posts.length : 0
});

export default connect(mapStateToProps, mapDispatchToProps)(HeaderNavBar);
