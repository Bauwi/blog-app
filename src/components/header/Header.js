import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Headroom from 'react-headroom';

import HeaderNavBar from './HeaderNavBar';

const Header = () => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/">
          <h1>_TEKKNE_</h1>
        </Link>
        <HeaderNavBar />
      </div>
    </div>
  </header>
);

export default Header;
