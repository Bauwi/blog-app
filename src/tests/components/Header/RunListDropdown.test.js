import React from 'react';
import { shallow, mount } from 'enzyme';
import { RunListDropdown } from '../../../components/header/RunListDropdown';
import run from '../../fixtures/run';
import { MemoryRouter } from 'react-router-dom';

test('should render RunListDropdown properly', () => {
  const wrapper = shallow(<RunListDropdown setCurrentPostRun={() => {}} startResetRun={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

// test('should call startResetRun on reset button click', () => {
//   const startResetRun = jest.fn();
//   const wrapper = shallow(<RunListDropdown posts={run.posts} setCurrentPostRun={() => {}} startResetRun={startResetRun} />);

//   wrapper
//     .find('button')
//     .at(1)
//     .simulate('click');

//   expect(startResetRun).toHaveBeenCalled();
// });

// test('calls "handleResetClick()" on button click', () => {
//   const wrapper = mount(<MemoryRouter>
//     <RunListDropdown posts={run.posts} setCurrentPostRun={() => {}} startResetRun={() => {}} />
//   </MemoryRouter>);
//   const spy = jest.spyOn(wrapper.instance(), 'handleResetClick');
//   wrapper.update();
//   wrapper
//     .find('button')
//     .at(1)
//     .simulate('click');
//   expect(spy).toHaveBeenCalled();
// });
