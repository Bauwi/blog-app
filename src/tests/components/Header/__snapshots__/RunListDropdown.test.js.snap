// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RunListDropdown properly 1`] = `
<Dropdown
  className="dropdown"
  mouseEnterDelay={0.15}
  mouseLeaveDelay={0.1}
  overlay={
    <Menu
      className=""
      prefixCls="ant-menu"
      theme="light"
    >
      <MenuItem>
        <button
          className="button"
          onClick={[Function]}
        >
          Resume
        </button>
        <button
          className="button"
          onClick={[Function]}
        >
          Reset
        </button>
      </MenuItem>
    </Menu>
  }
  placement="bottomCenter"
  prefixCls="ant-dropdown"
>
  <Link
    className="ant-dropdown-link"
    replace={false}
    to="/run"
  >
    <i
      className="fa fa-book"
    />
  </Link>
</Dropdown>
`;
